### PR TITLE
[inductor] More compile time optimizations

### PR DESCRIPTION
**Pull request by jansel**
Mostly improving `sympy` usage.  This give another ~2x improvement in warm compile time.

BERT_pytorch fp16 backwards before:
```
         25532823 function calls (24706453 primitive calls) in 14.294 seconds

   Ordered by: cumulative time
   List reduced from 5313 to 100 due to restriction <100>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   14.320   14.320 compile_fx.py:42(compile_fx_inner)
      6/3    0.000    0.000   14.288    4.763 utils.py:70(time_wrapper)
        1    0.000    0.000    9.977    9.977 graph.py:332(compile_to_fn)
        1    0.000    0.000    9.977    9.977 graph.py:315(compile_to_module)
        1    0.000    0.000    9.881    9.881 graph.py:307(codegen)
        1    0.004    0.004    6.572    6.572 scheduler.py:502(__init__)
     1768    0.015    0.000    6.229    0.004 ir.py:3289(__call__)
     2449    0.022    0.000    5.727    0.002 dependencies.py:249(extract_read_writes)
     1768    0.010    0.000    5.635    0.003 ir.py:3393(__call__)
      484    0.006    0.000    4.987    0.010 scheduler.py:220(__init__)
        1    0.000    0.000    4.009    4.009 graph.py:136(run)
        1    0.008    0.008    4.009    4.009 interpreter.py:95(run)
     3102    0.014    0.000    3.998    0.001 graph.py:294(run_node)
53758/53756    0.090    0.000    3.582    0.000 _print_helpers.py:27(__str__)
53758/53756    0.043    0.000    3.492    0.000 printer.py:371(__call__)
64669/50536    0.041    0.000    3.480    0.000 printer.py:290(doprint)
53758/53756    0.045    0.000    3.449    0.000 str.py:982(sstr)
124788/50536    0.480    0.000    3.448    0.000 printer.py:294(_print)
     6879    0.023    0.000    3.019    0.000 dependencies.py:195(load)
        1    0.004    0.004    2.985    2.985 scheduler.py:1012(codegen)
8652/8379    0.057    0.000    2.972    0.000 str.py:50(_print_Add)
      484    0.016    0.000    2.919    0.006 ir.py:1850(simplify_and_reorder)
      303    0.003    0.000    2.905    0.010 triton.py:1035(codegen_nodes)
      303    0.006    0.000    2.888    0.010 triton.py:1113(codegen_node_schedule)
    95418    0.938    0.000    2.796    0.000 basic.py:802(subs)
     3102    0.011    0.000    2.740    0.001 interpreter.py:144(run_node)
     4013    0.003    0.000    2.646    0.001 utils.py:161(wrapper)
     1198    0.009    0.000    2.587    0.002 ir.py:1781(get_read_writes)
     8881    0.014    0.000    2.535    0.000 expr.py:384(__format__)
     8881    0.006    0.000    2.503    0.000 {function Expr.__format__ at 0x7fe9b7aea670}
      968    0.003    0.000    2.371    0.002 ir.py:3226(__init__)
      968    0.016    0.000    2.367    0.002 ir.py:3310(__init__)
      484    0.006    0.000    2.296    0.005 scheduler.py:315(codegen)
8652/1216    0.016    0.000    2.283    0.002 lowering.py:272(inner_fn)
8652/1216    0.012    0.000    2.258    0.002 lowering.py:277(<listcomp>)
      910    0.003    0.000    1.947    0.002 ir.py:340(store_output)
     4272    0.007    0.000    1.895    0.000 sizevars.py:487(load)
     1768    0.008    0.000    1.859    0.001 ir.py:3388(make_gm)
     1768    0.035    0.000    1.819    0.001 graph_module.py:311(__init__)
     2711    0.006    0.000    1.761    0.001 graph.py:226(call_function)
7825/2711    0.031    0.000    1.752    0.001 lowering.py:154(wrapped)
     8652    0.018    0.000    1.725    0.000 printer.py:340(_as_ordered_terms)
20199/16663    0.040    0.000    1.706    0.000 module.py:1268(__setattr__)
8655/8653    0.055    0.000    1.705    0.000 expr.py:1113(as_ordered_terms)
```

After:
```
         15284480 function calls (14520161 primitive calls) in 7.152 seconds

   Ordered by: cumulative time
   List reduced from 1726 to 100 due to restriction <100>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    7.174    7.174 compile_fx.py:42(compile_fx_inner)
      6/3    0.000    0.000    7.143    2.381 utils.py:70(time_wrapper)
        1    0.000    0.000    5.102    5.102 graph.py:332(compile_to_fn)
        1    0.000    0.000    5.102    5.102 graph.py:315(compile_to_module)
        1    0.000    0.000    4.998    4.998 graph.py:307(codegen)
     1769    0.046    0.000    4.000    0.002 interpreter.py:95(run)
        1    0.003    0.003    3.481    3.481 scheduler.py:502(__init__)
    35434    0.096    0.000    3.406    0.000 interpreter.py:144(run_node)
      484    0.004    0.000    2.838    0.006 scheduler.py:220(__init__)
     1768    0.014    0.000    2.368    0.001 ir.py:3288(__call__)
     1768    0.006    0.000    2.264    0.001 ir.py:3388(__call__)
      484    0.013    0.000    2.169    0.004 ir.py:1849(simplify_and_reorder)
      968    0.004    0.000    1.890    0.002 ir.py:3225(__init__)
      968    0.013    0.000    1.887    0.002 ir.py:3310(__init__)
        1    0.000    0.000    1.769    1.769 graph.py:136(run)
     3102    0.012    0.000    1.759    0.001 graph.py:294(run_node)
     2449    0.019    0.000    1.632    0.001 dependencies.py:240(extract_read_writes)
    20973    0.026    0.000    1.559    0.000 interpreter.py:236(call_method)
        1    0.003    0.003    1.306    1.306 scheduler.py:1012(codegen)
      303    0.003    0.000    1.251    0.004 triton.py:1044(codegen_nodes)
      303    0.005    0.000    1.236    0.004 triton.py:1122(codegen_node_schedule)
    16048    0.039    0.000    1.193    0.000 proxy.py:52(create_proxy)
8652/1216    0.014    0.000    1.048    0.001 lowering.py:272(inner_fn)
8652/1216    0.010    0.000    1.025    0.001 lowering.py:277(<listcomp>)
     4272    0.006    0.000    0.990    0.000 sizevars.py:488(load)
     2711    0.005    0.000    0.920    0.000 graph.py:226(call_function)
7825/2711    0.029    0.000    0.912    0.000 lowering.py:154(wrapped)
    10292    0.009    0.000    0.880    0.000 proxy.py:355(__call__)
      484    0.005    0.000    0.837    0.002 scheduler.py:315(codegen)
    16048    0.010    0.000    0.786    0.000 proxy.py:32(create_node)
      910    0.003    0.000    0.779    0.001 ir.py:341(store_output)
    16048    0.042    0.000    0.776    0.000 graph.py:738(create_node)
   145136    0.080    0.000    0.769    0.000 node.py:595(map_arg)
     4013    0.003    0.000    0.718    0.000 utils.py:161(wrapper)
      768    0.003    0.000    0.698    0.001 ir.py:398(store_reduction)
     1198    0.007    0.000    0.688    0.001 ir.py:1778(get_read_writes)
315155/145136    0.429    0.000    0.677    0.000 node.py:603(map_aggregate)
     2935    0.007    0.000    0.608    0.000 ir.py:1702(loader)
     6879    0.016    0.000    0.605    0.000 dependencies.py:186(load)
        1    0.000    0.000    0.605    0.605 scheduler.py:727(fuse_nodes)
        2    0.002    0.001    0.604    0.302 scheduler.py:737(fuse_nodes_once)
43809/29684    0.026    0.000    0.592    0.000 printer.py:290(doprint)
52335/29684    0.208    0.000    0.574    0.000 printer.py:294(_print)
    32904    0.051    0.000    0.559    0.000 _print_helpers.py:27(__str__)
```

### CODE REVIEW

File: torchinductor/utils.py [hunk 1]
```diff
@@ -165,3 +165,50 @@ def wrapper(self):
...
+
+    def swap_key(key):
+        if isinstance(key, str):
+            return sympy.Symbol(key)
+        assert isinstance(key, sympy.Symbol), key
+        return key
+
+    def swap_value(val):
+        if isinstance(val, int):
+            return sympy.Integer(val)
```
Comment (asmeurer) [inline torchinductor/utils.py:205]: I'm curious why this is needed. I checked and it looks like `xreplace` automatically converts `int` to `Integer`.

File: torchinductor/utils.py [hunk 2]
```diff
@@ -165,3 +165,50 @@ def wrapper(self):
...
+def sympy_str(expr: sympy.Expr):
+    """
+    Normal sympy str is very slow, this is a lot faster.  The result are
+    somewhat worse, as it doesn't do as much simplification.  So don't
+    use this for final codegen.
+    """
+    if isinstance(expr, sympy.Symbol):
+        return expr.name
+    if isinstance(expr, sympy.Add):
+        return " + ".join(map(sympy_str, expr.args))
```
Comment (asmeurer) [inline torchinductor/utils.py:179]: You may be able to get the equivalent of this if you use `sympy.sstr(expr, order='none')`

Comment (asmeurer) [reply torchinductor/utils.py:179]: Which should make the rest of this function unnecessary as well.

File: torchinductor/utils.py (hunk 1 above)
Comment (jansel) [reply torchinductor/utils.py:205]: You are correct, it may not be needed.

File: torchinductor/utils.py (hunk 2 above)
Comment (jansel) [reply torchinductor/utils.py:179]: I tried replacing this with `return sympy.sstr(expr, order='none')`, overall ~39% slower:
```
         18643652 function calls (17885889 primitive calls) in 9.723 seconds

   Ordered by: cumulative time
   List reduced from 1726 to 100 due to restriction <100>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    9.748    9.748 compile_fx.py:42(compile_fx_inner)
      6/3    0.000    0.000    9.716    3.239 utils.py:70(time_wrapper)
     1769    0.050    0.000    6.351    0.004 interpreter.py:95(run)
        1    0.000    0.000    6.256    6.256 graph.py:332(compile_to_fn)
        1    0.000    0.000    6.256    6.256 graph.py:315(compile_to_module)
        1    0.000    0.000    6.157    6.157 graph.py:307(codegen)
    35434    0.104    0.000    5.145    0.000 interpreter.py:144(run_node)
        1    0.004    0.004    4.560    4.560 scheduler.py:502(__init__)
     2449    0.022    0.000    3.921    0.002 dependencies.py:240(extract_read_writes)
      484    0.006    0.000    3.567    0.007 scheduler.py:220(__init__)
     1768    0.016    0.000    3.324    0.002 ir.py:3288(__call__)
     1768    0.007    0.000    3.214    0.002 ir.py:3388(__call__)
        1    0.000    0.000    3.172    3.172 graph.py:136(run)
     3102    0.014    0.000    3.160    0.001 graph.py:294(run_node)
38586/24454    0.027    0.000    2.911    0.000 printer.py:290(doprint)
91842/24454    0.351    0.000    2.892    0.000 printer.py:294(_print)
27675/27674    0.024    0.000    2.818    0.000 printer.py:371(__call__)
27675/27674    0.026    0.000    2.794    0.000 str.py:982(sstr)
7411/7147    0.047    0.000    2.542    0.000 str.py:50(_print_Add)
     8877    0.012    0.000    2.473    0.000 utils.py:170(sympy_str)
    20973    0.028    0.000    2.465    0.000 interpreter.py:236(call_method)
     6879    0.020    0.000    2.338    0.000 dependencies.py:186(load)
      484    0.016    0.000    2.304    0.005 ir.py:1849(simplify_and_reorder)
     4013    0.003    0.000    2.143    0.001 utils.py:161(wrapper)
     1198    0.009    0.000    2.091    0.002 ir.py:1778(get_read_writes)
      968    0.003    0.000    1.995    0.002 ir.py:3225(__init__)
      968    0.015    0.000    1.991    0.002 ir.py:3310(__init__)
8652/1216    0.017    0.000    1.958    0.002 lowering.py:272(inner_fn)
8652/1216    0.012    0.000    1.934    0.002 lowering.py:277(<listcomp>)
      910    0.003    0.000    1.737    0.002 ir.py:341(store_output)
     7411    0.015    0.000    1.483    0.000 printer.py:340(_as_ordered_terms)
7414/7413    0.052    0.000    1.467    0.000 expr.py:1113(as_ordered_terms)
        1    0.003    0.003    1.386    1.386 scheduler.py:1012(codegen)
     2711    0.007    0.000    1.357    0.001 graph.py:226(call_function)
7825/2711    0.031    0.000    1.348    0.000 lowering.py:154(wrapped)
      303    0.003    0.000    1.329    0.004 triton.py:1044(codegen_nodes)
      303    0.006    0.000    1.313    0.004 triton.py:1122(codegen_node_schedule)
     2935    0.009    0.000    1.265    0.000 ir.py:1702(loader)
7414/7413    0.144    0.000    1.263    0.000 expr.py:1167(as_terms)
    16048    0.038    0.000    1.244    0.000 proxy.py:52(create_proxy)
      768    0.003    0.000    1.027    0.001 ir.py:398(store_reduction)
     2755    0.009    0.000    1.024    0.000 ir.py:1288(loader)
     4272    0.006    0.000    1.017    0.000 sizevars.py:488(load)
      582    0.003    0.000    0.983    0.002 ir.py:3199(num_reads)
```

File: torchinductor/dependencies.py [hunk 3]
```diff
@@ -44,9 +46,9 @@ def maybe_swap_sizes(self) -> "MemoryDep":
         ):
             c = canonicalization_prefix()
             size = (self.size[1], self.size[0])
-            s0 = sympy.Symbol(c + "0", is_integer=True)
```
Comment (Chillee) [inline torchinductor/dependencies.py:47]: Does `is_integer` slow it down?

**Review by Chillee — APPROVED** [review]

File: torchinductor/dependencies.py (hunk 3 above)
Comment (jansel) [reply torchinductor/dependencies.py:47]: No, but:
```
>>> sympy.Symbol("x") == sympy.Symbol("x")
True
>>> sympy.Symbol("x") == sympy.Symbol("x", is_integer=True)
False
```

Comment (Chillee) [reply torchinductor/dependencies.py:47]: 🤔 Doesn't this indicate that we should just be setting `is_integer` to be True or False everywhere?

I'm worried if not setting `is_integer` prohibits certain simplifications.


For ex:

(Also, shouldn't we be setting `integer=True` and not `is_integer=True`? cc: @asmeurer )

```
>>> a = sympy.Symbol('a',positive=True)
>>> a * 2 > 1
2*a > 1
>>> a = sympy.Symbol('a', is_integer=True, positive=True)
>>> a * 2 > 1
2*a > 1
>>> a = sympy.Symbol('a', integer=True, positive=True)
>>> a * 2 > 1
True
```

Comment (jansel) [reply torchinductor/dependencies.py:47]: Yeah, I think we need to set it uniformly everywhere.

Currently, we make no usage of sympy inequalities, so I don't think this setting actually matters.

**Pull request merged**